Add prefferedDuringScheduling node affinity support for TAS (#10903)
kubernetes-sigs/kueue: keps/2724-topology-aware-scheduling/README.md

We propose to remove the Annotation and replace it with new field in Workload status that will it will contain
a list of failed nodes:

> ```golangS

Reviewer: typo to add "S"?

Author: removed

| [[10, 5], [5, 5, 5]] | 15|1 |  [[0, 0], [5, 5, 5]] | prefer more balanced rack
| [[15],[15]] [[15, 15]] | 25 | 1 | [[0],[0]] [[13, 12]] | prefer block where the request fits in a single rack
| [[15], [15], [15, 15]] | 25|5 |  [[0], [0], [15, 10]] | `podset-slice-required-topology = hostname`

### Support for Preferred Node Affinity

TAS supports evaluating `preferredDuringSchedulingIgnoredDuringExecution`
node affinities during the scheduling cycle. When the
`TASPreferredSchedulingAffinity` feature gate is enabled, the preferred
affinity scores are evaluated when selecting candidate topology domains,
allowing affinity preferences to guide placement decisions. Preferred affinity

> takes precedence over capacity fitness of a domain.

Reviewer: ```suggestion
In 0.18 we introduce the pilot support for domain affinity based on `preferredDuringSchedulingIgnoredDuringExecution`, behind the 
`TASRespectNodeAffinityPreferred` feature gate (alpha).

When the feature gate is enabled, we compute the "domain affinity scores" at all levels by
aggregating (summing) the scores from child topology domains.

Then, the "domain affinity scores" takes precedence over standard placement ordering of domains (based on BestFit or LeastFreeCapacity policies).

This feature only works when the lowest topology level specified in the Topology CRD is `kubernetes.io/hostname`.
```
To include the key informations:
- propose to rename as TASRespectNodeAffinityPreferred to reflect we are "respecting" api which is already there in API
- the version when the new semantic is introduced
- the mechanism of aggregation and evaluation in a bit more detail so that someone can imagine impl just by text
- make it clear the feature only works when "hostname" is the lowest level.

Author: done

> 
> #### Future Vision
> The long-term goal is to decommission this strict precedence model and its
> associated `TASPreferredSchedulingAffinity` feature gate in favor of a generic
> weighted scoring model of domains. Similar to the kube-scheduler scheduling
> plugins scoring model, candidate topology domains will be evaluated using
> a configurable list of individual scoring components (such as affinity
> preferences and capacity packing efficiency) that contribute to a normalized,
> unified final score to determine the globally optimal placement.

Reviewer: ```suggestion
#### Future Vision
One drawback of the proposed solution is that it introduces a strict precedence between the 
"domain affinity score" and the "placement policy". 

A more generic approach using weighted scoring model of domains is considered, similar to
the kube-scheduler scheduling plugins scoring model, candidate topology domains will be
evaluated using a configurable list of individual scoring components (such as affinity
preferences and capacity packing efficiency) that contribute to a normalized,
unified final score to determine the globally optimal placement.

We are going to re-evaluate the next steps based on the user and community feedback.
```

Author: done

- perform full scheduling simulation rather than just capacity counting
 (including pod affinities and anti-affinities)
- drop `TASLeastAllocated` feature gate
- drop `TASPreferredSchedulingAffinity` feature gate upon decommissioning the
 strict precedence affinity model in favor of the Weighted Scoring Model of

>  domains (see [Future Vision](#future-vision))

Reviewer: ```suggestion
- re-evaluate the status of the `TASPreferredSchedulingAffinity` feature gate - either graduate, or decommission with the more generic model of weighted scores (see [Future Vision](#future-vision))
```

Author: done
